x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java

```diff
@@ -605,7 +605,18 @@ public Expression visitRegexBooleanExpression(EsqlBaseParser.RegexBooleanExpress
         Literal pattern = visitString(ctx.pattern);
         RegexMatch<?> result = switch (type) {
             case EsqlBaseParser.LIKE -> new WildcardLike(source, left, new WildcardPattern(pattern.fold().toString()));
-            case EsqlBaseParser.RLIKE -> new RLike(source, left, new RLikePattern(pattern.fold().toString()));
+            case EsqlBaseParser.RLIKE -> {
+                var patternString = pattern.fold().toString();
+                var rlike = new RLike(source, left, new RLikePattern(patternString));
+                try {
+                    // the pattern could be invalid
+                    // this forces to build the automaton, so that in case it will fail right away with a meaningful error message
+                    rlike.pattern().matchesAll();
```

Comment:
Question/Thought: Maybe sometimes building the automaton could be expensive, and in rare cases even irrelevant (when RLIKE get's optimized away); maybe we can move the validation a bit later, but probably this is just fine the way it is :)

I also wonder if the memory taken up by the automata of many RLIKEs can be significant; we already know that our query plans sometimes take up quite some memory which needs to be tracked.

Reply:
> maybe we can move the validation a bit later, but probably this is just fine the way it is :)

I put it in the parser just because I consider it a syntax error (for the same reason I throw a ParsingException, not a ValidationException). 
I don't have a very strong opinion though, I could move it it to the Verifier, but it wouldn't make a big difference in terms of cost (the optimizations didn't happen yet). 
Doing validation in the logical optimizer seems a bit late though, and I'm not sure I'd manage to do it before pruning the possibly redundant RLIKEs.

About memory management, thanks for pointing it out. I don't have an answer unfortunately, but I'll investigate it,.

Reply:
Imho, I don't think the additional `Invalid regex pattern for RLIKE` message it's worth the performance penalty of **always** trying to resolve the regex pattern. @luigidellaquila before this change here where is the first place where we try to build the automaton?

Reply:
It's during logical plan optimization, in [ReplaceRegexMatch](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/ReplaceRegexMatch.java#L28) rule.

Most of the time the pattern will be resolved anyway, so unless the RLIKE is completely pruned, the performance will practically be the same.

I'd really like to have a ParsingException, to point to the exact position where the problem happened, but throwing it in the optimizer doesn't seem correct...

Reply:
> Most of the time the pattern will be resolved anyway, so unless the RLIKE is completely pruned, the performance will practically be the same.

It's not only that. After the parser `Builder`s, there is the `IndexResolver` (and its `field_caps` call and index names validation), there is the `Analyzer` and the `Verifier` (where the execution can be stopped if field names are not right or invalid data types are used, for example) where the execution can fail. And at those points the automaton has already been built and checked needlessly.

My suggestions (up for discussion):
- throw the `ParsingException` in the `ReplaceRegexMatch` rule where you catch the IAE from the automaton creation and checks for `matchesAll` **and** `exactMatch` (a pattern can fail here as well? not only from `matchesAll`). At that point you still have the line numbers and the pattern itself (`throw new ParsingException(regexMatch.source(), "Invalid regex pattern for RLIKE [{}]: [{}]", pattern.asJavaRegex(), iae.getMessage());`). The only drawback here is the apparently wrong location to throw a ParsingException from; imho, it's worth it though.
- think about the error message itself a bit more, because `org.apache.lucene.util.automaton.Operations` throws the IAE exception in other cases as well. Even though, "Invalid regex pattern for RLIKE" is generic enough I believe...

Reply:
I slightly lean towards throwing the exception a bit later (when the pattern is first used), too, as long as the error message is the same/just as useful to the user. But on my end, that's mostly paranoia from unforeseen consequences of eagerly building the automaton; I can't think of cases off the top of my head where being eager is very bad (except that we could sometimes fail a query without even building the automaton, as Andrei pointed out).

Maybe we could move the try/catch simply into `RLikePattern.createAutomaton()`? This would guarantee we remain as lazy as we currently are.

Reply:
Thanks for the feedback folks!
Again, I don't have a very strong opinion on this, I'm good with moving the exception check to the optimizer 👍 
`RLikePattern.createAutomaton()` is the place that throws the exception, but yeah, it would require the Source, not sure it's worth it.
Practically, a check in `ReplaceRegexMatch` should solve the problem for now, but it could be a bit fragile in the long term, eg. if we introduce another optimization rule that builds the automaton.

Reply:
Also, if it's too much trouble or it's more involved for `like`, create a separate issue for it and deal with it separately.